<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Curso-de-Python-Intermedio" data-toc-modified-id="Curso-de-Python-Intermedio-0">Curso de Python Intermedio</a></span><ul class="toc-item"><li><span><a href="#Módulo-1.-Preparación" data-toc-modified-id="Módulo-1.-Preparación-0.1">Módulo 1. Preparación</a></span><ul class="toc-item"><li><span><a href="#El-Zen-de-Python" data-toc-modified-id="El-Zen-de-Python-0.1.1">El Zen de Python</a></span></li><li><span><a href="#Documentación" data-toc-modified-id="Documentación-0.1.2">Documentación</a></span></li></ul></li><li><span><a href="#Módulo-2.-Entorno-virtual" data-toc-modified-id="Módulo-2.-Entorno-virtual-0.2">Módulo 2. Entorno virtual</a></span><ul class="toc-item"><li><span><a href="#¿Qué-es-un-entorno-virtual?" data-toc-modified-id="¿Qué-es-un-entorno-virtual?-0.2.1">¿Qué es un entorno virtual?</a></span></li><li><span><a href="#Creación-de-un-entorno-virtual" data-toc-modified-id="Creación-de-un-entorno-virtual-0.2.2">Creación de un entorno virtual</a></span></li><li><span><a href="#Instalación-de-dependencias-con-pip" data-toc-modified-id="Instalación-de-dependencias-con-pip-0.2.3">Instalación de dependencias con pip</a></span></li><li><span><a href="#Una-alternativa:-Anaconda" data-toc-modified-id="Una-alternativa:-Anaconda-0.2.4">Una alternativa: Anaconda</a></span></li></ul></li><li><span><a href="#Módulo-3.-Alternativa-a-los-ciclos:-comprehensions" data-toc-modified-id="Módulo-3.-Alternativa-a-los-ciclos:-comprehensions-0.3">Módulo 3. Alternativa a los ciclos: comprehensions</a></span><ul class="toc-item"><li><span><a href="#Listas-y-diccionarios-anidados" data-toc-modified-id="Listas-y-diccionarios-anidados-0.3.1">Listas y diccionarios anidados</a></span><ul class="toc-item"><li><span><a href="#Listas:" data-toc-modified-id="Listas:-0.3.1.1">Listas:</a></span></li><li><span><a href="#Tuplas-y-secuencias:" data-toc-modified-id="Tuplas-y-secuencias:-0.3.1.2">Tuplas y secuencias:</a></span></li><li><span><a href="#Diccionarios:" data-toc-modified-id="Diccionarios:-0.3.1.3">Diccionarios:</a></span></li></ul></li><li><span><a href="#List-Comprehensions" data-toc-modified-id="List-Comprehensions-0.3.2">List Comprehensions</a></span></li><li><span><a href="#Dictionary-comprehensions" data-toc-modified-id="Dictionary-comprehensions-0.3.3">Dictionary comprehensions</a></span></li></ul></li><li><span><a href="#Módulo-4.-Conceptos-avanzados-de-funciones" data-toc-modified-id="Módulo-4.-Conceptos-avanzados-de-funciones-0.4">Módulo 4. Conceptos avanzados de funciones</a></span><ul class="toc-item"><li><span><a href="#Funciones-anónimas:-lambda" data-toc-modified-id="Funciones-anónimas:-lambda-0.4.1">Funciones anónimas: lambda</a></span></li><li><span><a href="#High-order-functions:-filter,-map-y-reduce" data-toc-modified-id="High-order-functions:-filter,-map-y-reduce-0.4.2">High order functions: filter, map y reduce</a></span><ul class="toc-item"><li><span><a href="#filter():" data-toc-modified-id="filter():-0.4.2.1"><code>filter()</code>:</a></span></li><li><span><a href="#map():" data-toc-modified-id="map():-0.4.2.2"><code>map()</code>:</a></span></li><li><span><a href="#reduce():" data-toc-modified-id="reduce():-0.4.2.3"><code>reduce()</code>:</a></span></li></ul></li><li><span><a href="#Proyecto:-Filtrando-datos" data-toc-modified-id="Proyecto:-Filtrando-datos-0.4.3">Proyecto: Filtrando datos</a></span></li></ul></li><li><span><a href="#Módulo-5.-Manejo-de-errores" data-toc-modified-id="Módulo-5.-Manejo-de-errores-0.5">Módulo 5. Manejo de errores</a></span><ul class="toc-item"><li><span><a href="#Los-errores-en-el-código" data-toc-modified-id="Los-errores-en-el-código-0.5.1">Los errores en el código</a></span><ul class="toc-item"><li><span><a href="#Errores-de-sintaxis" data-toc-modified-id="Errores-de-sintaxis-0.5.1.1">Errores de sintaxis</a></span></li><li><span><a href="#Excepciones" data-toc-modified-id="Excepciones-0.5.1.2">Excepciones</a></span></li><li><span><a href="#Traceback" data-toc-modified-id="Traceback-0.5.1.3">Traceback</a></span></li></ul></li><li><span><a href="#Debugging-o-depuración" data-toc-modified-id="Debugging-o-depuración-0.5.2">Debugging o depuración</a></span><ul class="toc-item"><li><span><a href="#BreakPoints" data-toc-modified-id="BreakPoints-0.5.2.1">BreakPoints</a></span></li></ul></li><li><span><a href="#Manejo-de-excepciones" data-toc-modified-id="Manejo-de-excepciones-0.5.3">Manejo de excepciones</a></span><ul class="toc-item"><li><span><a href="#Try-/-Except:" data-toc-modified-id="Try-/-Except:-0.5.3.1"><code>Try / Except</code>:</a></span></li><li><span><a href="#raise:" data-toc-modified-id="raise:-0.5.3.2"><code>raise</code>:</a></span></li><li><span><a href="#finally:" data-toc-modified-id="finally:-0.5.3.3"><code>finally</code>:</a></span></li></ul></li><li><span><a href="#Poniendo-a-prueba-el-manejo-de-excepciones" data-toc-modified-id="Poniendo-a-prueba-el-manejo-de-excepciones-0.5.4">Poniendo a prueba el manejo de excepciones</a></span></li><li><span><a href="#Assert-statements" data-toc-modified-id="Assert-statements-0.5.5">Assert statements</a></span></li></ul></li><li><span><a href="#Módulo-6.-Manejo-de-archivos" data-toc-modified-id="Módulo-6.-Manejo-de-archivos-0.6">Módulo 6. Manejo de archivos</a></span><ul class="toc-item"><li><span><a href="#¿Cómo-trabajar-con-archivos?" data-toc-modified-id="¿Cómo-trabajar-con-archivos?-0.6.1">¿Cómo trabajar con archivos?</a></span></li><li><span><a href="#Trabajando-con-archivos-de-texto-en-Python" data-toc-modified-id="Trabajando-con-archivos-de-texto-en-Python-0.6.2">Trabajando con archivos de texto en Python</a></span></li></ul></li><li><span><a href="#Conclusiones." data-toc-modified-id="Conclusiones.-0.7">Conclusiones.</a></span><ul class="toc-item"><li><span><a href="#Programa-final:-Hangman-Game" data-toc-modified-id="Programa-final:-Hangman-Game-0.7.1">Programa final: Hangman Game</a></span><ul class="toc-item"><li><span><a href="#Ayudas-y-pistas:" data-toc-modified-id="Ayudas-y-pistas:-0.7.1.1">Ayudas y pistas:</a></span></li><li><span><a href="#Posibles-mejoras-al-juego:" data-toc-modified-id="Posibles-mejoras-al-juego:-0.7.1.2">Posibles mejoras al juego:</a></span></li><li><span><a href="#Código:" data-toc-modified-id="Código:-0.7.1.3">Código:</a></span></li></ul></li></ul></li></ul></li></ul></div>

---

# Curso de Python Intermedio

## Módulo 1. Preparación

### El Zen de Python

Es una colección de 20 principios de software que influyen en el diseño del lenguaje, de los cuales 19 fueron escritos por Tim Peters. 

Como easter egg podemos acceder a este Zen de Python desde la terminal con el comando:

In [44]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


---

### Documentación

Documentar el código de un programa es añadir suficiente información como para explicar lo que hace, punto por punto, de forma que no sólo los ordenadores sepan qué hacer, sino que además los humanos entiendan qué están haciendo y por qué.

Documentar un programa no es sólo una buena práctica de un programador, si no que además, es una necesidad que sólo se aprecia en su debida magnitud cuando hay errores que reparar o hay que entender el programa con nuevas capacidades o adaptarlo a un nuevo escenario. Hay 2 cosas que hay que tener en mente:

1. Todos los programas tienen errores y descubrirlos es sólo cuestión de tiempo y de que el programa tenga éxito y se utilice frecuentemente o por una gran cantidad de usuarios.
    
    
2. Todos los programas sufren modificaciones a lo largo de su vida, al menos todos aquellos que tienen éxito.

En la documentación de Python podemos encontrar todas las novedades que se agregan en cada actualización, tutoriales para entender cómo funciona cada una de sus funciones y en general todo el lenguaje en su nivel más fundamental y con más detalle, descripciones de la sintaxis y los elementos del lenguaje entre muchas otras cosas.

[Python 3.9.4 documentation](https://docs.python.org/3/)

También tenemos algo llamado PEP o Propuestas de Mejora de Python, que es una guía de estilo adoptada ampliamente por la comunidad de Python. Facilita la lectura del código y hace más homogéneo el código entre diferentes programadores.

Son documentos que nos dicen como funciona el lenguaje y como deberíamos escribir nuestros códigos de manera correcta. Entre todos los PEP destaca el PEP8 ya que es una guía que nos dice como deberíamos de escribir código, buenas prácticas, como seleccionar los nombres de clase, como trabajar la indentación, etc.

[PEP 8 -- Style Guide for Python Code](https://www.python.org/dev/peps/pep-0008/)

[Fragmento de la documentación PDF](https://drive.google.com/file/d/1dSYgA7Xonw1XLMcjrsMXSqDUWLlqi-57/view)

---

## Módulo 2. Entorno virtual

### ¿Qué es un entorno virtual?

Primero, recordemos el concepto de módulo. Un módulo es un fragmento de código escrito por otra persona que a nosotros nos sirve para resolver determinado problema. De forma que no tengamos que escribir de nuevo todo ese código por nuestra cuenta y en su lugar, aprovechemos de herramientas que ya han sido creadas y compartidas por otros usuarios.

Por ejemplo, anteriormente ya hemos importado el módulo `random` que nos permite trabajar con números aleatorios.

![](https://i.imgur.com/Mcd2izU.png)

Ahora hablemos de ambientes o entornos virtuales, imaginemos primero a nuestra computadora que ya tiene instalado Python y tenemos varios proyectos en los que estamos trabajando. Dependiendo del problema en el que estamos trabajando en cada proyecto nos puede interesar tener varios módulos con los que trabajar, por ejemplo, random, herramientas de gráficado, etc. 

El problema es que si tenemos todo instalado en nuestro ambiente local (principal) puede que en un momento todo funcione correctamente, pero hay veces donde con las actualizaciones de ciertos módulos podamos tener ciertos problemas. Por ejemplo, hay ocasiones donde el Módulo 2 se actualiza y este tiene problema o deja de ser compatible con el Módulo 1, por lo que nos trae problemas en todos nuestros proyectos al tenerlos instalados simultáneamente en todo nuestro ambiente. 

Puede que en el proyecto 1 solo nos interesa tener instalado el módulo 2 y en el proyecto 2 el módulo 1, por lo que no tendría que haber problemas de incompatibilidad si los tuvieramos instalados por separado. De aquí es donde sale la necesidad de los *entornos virtuales*.

Lo que queremos lograr con los ambientes virtuales es, por así decirlo, dividir o seccionar nuestro computador en varias partes donde cada parte funciona correctamente con Python y los módulos que queremos usar en cada proyecto.

![](https://i.imgur.com/RANUzbx.png)

Podemos imaginarlo como si estuvieramos clonando el lenguaje de programación en "varios lenguajes independientes", que más bien serían extensiones del lenguaje. Por lo que al momento de actualizar los módulos, nosotros podemos elegir en que proyecto queremos actualizarlos y en cuales no para evitar problemas de compatibilidad.

![](https://i.imgur.com/wSQ3eto.png)

---

### Creación de un entorno virtual

Para empezar a crear nuestro ambiente virtual vamos a realizar los siguientes pasos:

1. Crear una nueva carpeta destinada al proyecto (con `mkdir nombre_de_la_carpeta` desde la terminal).


2. Nos posicionamos en la carpeta que creamos (`cd nombre_de_la_carpeta`)


3. Como buena práctica inicializamos a la carpeta como un repositorio de Git, para mantener un control de las versiones (`git init`).  Esto nos posicionará en la rama *Master* o *Main* dependiendo la versión y configuración de Git.


4. Creamos un "Python nuevo", que es el que va a correr en nuestro proyecto 

    * `py -m venv nombre_del_entorno_virtual` para windows 
    * `python3 -m venv nombre_del_entorno_virtual` para un sistema de linux. 
    
    Donde `-m` hace referencia a que llamaremos a un módulo del lenguaje y `venv` significa *Virtual Enviorement*. Y por convención el nombre del ambiente suele ser `venv` o `env` . De forma que el comando completo con la convención nos quedaría así:
    
     `py -m venv venv`
     
     `py -m venv env`
     
     
5. Una vez que ya está creado nuestro ambiente virtual verificamos que este correctamente creado con el comando `ls` que nos muestra los archivos y carpetas para posteriormente, activar nuestro ambiente virtual. El comando que nos permite activar el ambiente virtual se encuentra en la carpeta `Scripts/` dentro de `venv/`, por lo que para activarlo tendremos que usar la siguiente instrucción:
    
    * Windows: `.\venv\Scripts\activate` o `.\venv\Scripts\activate.bat`
    
    * Linux/Mac: `source venv/bin/activate`
    
    Para salir del ambiente basta con usar el comando:
    
    `deactivate`
    
    Para que no sea tan tedioso escribir toda la ruta al activar nuestro ambiente virtual podemos crear un *alias* para acortar el tiempo en el que entramos, en Windows se crea de la siguiente forma:
    
    `alias nombre_del_alias=.\venv\Scripts\activate.bat`
    
    `avenv=.\venv\Scripts\activate.bat`
    
    Y para acceder solo tendríamos que hacer una llamada al comando
    
    `avenv`

Para el caso de sistemas Linux o Mac también podemos usar el alias, pero una vez salimos de la terminal se borrará, por lo que no es de mucha utilidad.

---

### Instalación de dependencias con pip

Dentro de Python tenemos incorporados algunos módulos que por así decirlo, ya vienen incorporados "de fabrica", por lo que no tenemos que realizar ninguna instalación extra. Otra historia muy diferente es cuando queremos utilizar ciertos módulos o librerías como *numpy, qiskit, pandas, Pytest, etc*. Los cuales si bien están dentro de Python, no vienen incorporados desde el inicio en nuestro sistema, así que tendremos que realizar una instalación.

Para realizar la instalación tenemos que hacer uso de un manejador o instalador de dependencias, en este caso utilizaremos PIP (Package Installer for Python). Básicamente es un módulo que viene incorporado "de fabrica" que nos permite instalar de forma más cómoda todos esos módulos extras que queremos incorporar a Python.

**NOTA:** Como buena práctica, deberíamos utilizar PIP únicamente dentro de nuestro ambiente virtual. Al menos de que sea un paquete que estamos seguros de que utilizaremos en cada uno de nuestros proyectos, lo cual es algo poco común.

Para realizar nuestras instalaciones de paquetes con PIP debemos de seguir los siguientes pasos:

1. Activar nuestro entorno virtual, para que los paquetes solo se instalen ahí y no tengamos problemas futuros con sus actualizaciones.


2. Podemos verificar los paquetes que tenemos instalados usando el comando: 

    `pip freeze` en windows 
    
    `pip3` en algunos casos de linux o Mac
    
    
3. Realizamos la instalación de la siguiente forma, en este caso instalaremos el módulo Pandas como ejemplo:

    `pip install pandas`
    
    En el caso de Pandas podremos notar que se instalan diferentes paquetes, esto es por que Pandas es un módulo muy complejo y depende de otros paquetes. Al realizar la instalación de Pandas ya incluirá el resto de paquetes que necesite para funcionar por si mismo. Si hacemos de nuevo el comando `pip freeze` veremos que ahora tenemos instalados los siguientes módulos:
    
    * numpy==1.20.2 
    * pandas==1.2.4 
    * python-dateutil==2.8.1 
    * pytz==2021.1 
    * six==1.15.0
    
   En JavaScript teníamos algo muy similar a `pip` al utilizar el comando `npm`.
    
    
4. Una vez realizada la instalación podremos hacer uso de dicho módulo en todos los proyectos que creemos dentro de esa carpeta.

**NOTA:** En caso de ser necesario, podemos actualizar PIP con el comando:

`py -m pip install --upgrade pip`

Ahora, **¿Que pasaría cuando queremos compartir un proyecto con alguien más y estamos haciendo uso de muchos módulos extras que tuvimos que instalar independientemente?** Para que esa persona pueda utilizar tu programa tendría que tener instaladas las mismas dependencias y paquetes que tú, las mismas versiones, etc.

Para arreglar eso podemos compartir nuestro archivo de dependencias con el comando:

`pip freeze > requirements.txt`

Lo que hace este comando es guardar las dependencias con sus respectivas versiones para poder compartir el proyecto con otra persona. 

Una vez hecho esto, tendremos en nuestra carpeta el archivo `requirements.txt`. Si estamos en consola podemos abrir este archivo con el comando `cat`.

`cat requirements.txt`

y nos aparecen las dependencias que guardamos de nuestro proyecto.

Una vez hemos compartido nuestro proyecto, por ejemplo en GitHub y alguien más lo ha clonado en su propio equipo local. Lo único que tiene que hacer para utilizar tu proyecto con exactamente las mismas dependencias es:

1. Crear y activar su entorno virtual


2. Usar el comando:

    `pip install -r requirements.txt`
    
---

### Una alternativa: Anaconda

Anaconda es una distribución especial de Python, es un software pensado para el Data Science. Es uns distribución que nos permite crear ambientes virtuales e instalar dependencias de una manera gráfica. Es muy buena opción para ciencia de datos, pero para otras aplicaciones como backend es mejor utilizar pip u otros instaladores de paquetes.

Para instalar Anaconda tenemos que ir al [sitio oficial](https://www.anaconda.com/products/individual) donde podemos descargar los instaladores gráficos de Anaconda, excepto si estamos utilizando Linux ya que ahí tendremos que instalarlo desde la consola. 

Una vez descargado, ejecutamos y nos va abrir un instalador de windows de toda la vida, una vez llegando a la siguiente parte es importante que desmarquemos estas 2 opciones ya que pueden afectar el funcionamiento de Python:

![](https://i.imgur.com/aDNK1vv.png)

Una vez esta instalado podemos acceder desde nuestras aplicaciones a `Anaconda Navigator`

Al abrirlo tendremos una ventana como la siguiente:

![](https://i.imgur.com/DaRHFEx.png)

Y podemos crear ambientes virtuales de forma gráfica en la siguiente pestaña:

![](https://i.imgur.com/UHMIDfY.png)

A diferencia de PIP, con Anaconda ya tendremos instaladas varias dependencias de base que son utilizadas comúnmente en la ciencia de datos.

---

## Módulo 3. Alternativa a los ciclos: comprehensions

### Listas y diccionarios anidados

**NOTA:** Cuando creamos un ambiente virtual en una carpeta que posteriormente subiremos a Git y GitHub tendremos que crear un archivo `.gitignore` en nuestra carpeta donde haremos el ambiente virtual para no llevarnos los archivos del ambiente virtual al repositorio.

Una vez creamos este archivo escribimos `venv/`, de está manera ignorará todos los datos que tengamos en esa carpeta.

Ahora, hablemos un poco sobre estructuras de datos. Entre ellas tenemos listas, diccionarios y tuplas; *recordemos rápidamente cada una de ellas*.

#### Listas:

Python tiene varios tipos de datos compuestos, utilizados para agrupar otros valores. El más versátil es la *lista*, la cual puede ser escrita como una lista de valores separados por coma (items) entre corchetes. Las listas pueden contener ítems de diferentes tipos, pero usualmente los ítems son del mismo tipo.

In [45]:
lista = [1, 34, 5435, 6, 878, 989, 9]

al igual que las cadenas (y todas las demás tipos de integrados sequence), las listas se pueden indexar y segmentar.

In [46]:
lista[0]

1

In [47]:
lista[-1]

9

In [48]:
lista[-3]

878

Todas las operaciones de rebanado retornan una nueva lista que contiene los elementos pedidos. Esto significa que la siguiente rebanada retorna una shallow copy de la lista.

In [49]:
lista[:]

[1, 34, 5435, 6, 878, 989, 9]

También admiten operaciones como concatenación:

In [50]:
lista + [36, 49, 54, 846, 200]

[1, 34, 5435, 6, 878, 989, 9, 36, 49, 54, 846, 200]

A diferencia de las cadenas, que son *inmutables*, las listas son de tipo *mutable*, es decir, pueden cambiar su contenido:

In [51]:
lista2 = [1,2,3,4,5]

In [52]:
lista2[3] = 600

In [53]:
lista2

[1, 2, 3, 600, 5]

También puede agregar nuevos elementos al final de la lista, utilizando el método `append()`:

In [54]:
lista2.append(2500)

In [55]:
lista2.append(5*5*5*5)

In [56]:
lista2

[1, 2, 3, 600, 5, 2500, 625]

También es posible asignar una rebanada, y esto incluso puede cambiar la longitud de la lista o vaciarla completamente:

In [57]:
lista3 = ['a','b','c','d','e','f','g']

In [58]:
lista3

['a', 'b', 'c', 'd', 'e', 'f', 'g']

In [59]:
#Reemplazamos algunos de los valores
lista3[2:5] = ['C','D','E']

In [60]:
lista3

['a', 'b', 'C', 'D', 'E', 'f', 'g']

In [61]:
#Removemos un item 
lista3[2:5] = []

In [62]:
lista3

['a', 'b', 'f', 'g']

In [63]:
#Limpiamos la lista reemplazando TODOS los elementos por una lista vacía
lista3[:] = []

In [64]:
lista3

[]

La función predefinida `len()` también sirve para las listas (devuelve la longitud):

In [65]:
lista4 = ['a','sas','sadsdas','trwetre','dasag']

In [66]:
len(lista4)

5

Es posible anidar listar (crear listas que contengan otras listas), por ejemplo:

In [67]:
lista5 = ['a', 'b', 'c']
lista6 = [1,2,3]
lista7 = [lista5, lista6]

In [68]:
lista7

[['a', 'b', 'c'], [1, 2, 3]]

#### Tuplas y secuencias:

La listas y cadenas tienen propiedades en común, como el indizado y las operaciones de seccionado. Estas son dos ejemplos de datos de tipo *secuencial*. Existe otro tipo de datos de tipo *secuencial* estándar: la **tupla**.

Una tupla cosiste en un número de valores separados por comas, por ejemplo:

In [69]:
t1 = 12345, 54321, "oli" 

In [70]:
t1[0]

12345

In [71]:
t1

(12345, 54321, 'oli')

Podemos anidas tuplas

In [72]:
t2 = t1, (1,2,3,4,5)

In [73]:
t2

((12345, 54321, 'oli'), (1, 2, 3, 4, 5))

Las tuplas son **inmutables**

In [74]:
t1[0] = 88888

TypeError: 'tuple' object does not support item assignment

Pero **si pueden contener objetos mutables**:

In [75]:
t3 = ([1,2,3], [3,2,1])

In [76]:
t3

([1, 2, 3], [3, 2, 1])

Las tuplas siempre **se encierran en paréntesis**, para que las tuplas anidadas puedan interpretarse correctamente; pueden ingresarse con o sin paréntesis, aunque a menudo los paréntesis son necesarios de todas formas (si la tupla es parte de una expresión más grande). No es posible asignar a los ítems individuales de una tupla, pero sin embargo sí se puede crear tuplas que contengan objetos mutables, como las listas.

A pesar de que las tuplas puedan parecerse a las listas, frecuentemente se utilizan en distintas situaciones y para distintos propósitos. Las tuplas son **inmutables** y normalmente contienen una secuencia heterogénea de elementos que son accedidos al desempaquetar o indizar (o incluso acceder por atributo en el caso de las `namedtuples`). Por otro lado, las listas son **mutables**, y sus elementos son normalmente homogéneos y se acceden iterando la lista.

Un problema particular es la construcción de tuplas que contengan 0 o 1 ítem: la sintaxis presenta algunas peculiaridades para estos casos. Las tuplas vacías se construyen mediante un par de paréntesis vacío; una tupla con un ítem se construye poniendo una coma a continuación de un valor (no alcanza con encerrar un único valor entre paréntesis). Por ejemplo:

In [77]:
empty = ()

In [78]:
tupla_unica = 'hello', # Recordemos usar la coma

In [79]:
len(empty)

0

In [80]:
len(tupla_unica)

1

In [81]:
tupla_unica

('hello',)

En nuestras declaraciones del tipo 

In [82]:
t4 = 12345, 54321, 'holi'

es un ejemplo de *empaquetadp de tuplas*: los valores `12345`, `54321` y `holi` se empaquetan juntos en una tupla. La operación inversa también es posible y es la siguiente: `x, y, z = t`.

Esto se llama, **desempaquetado de secuencias**, y funciona para cualquier secuencia en el lado derecho del igual. **El desempaquetado de secuencias requiere que la cantidad de variables a la izquierda del signo igual sea del tamaño de la secuencia**. Notemos que la asignación múltiple es en realidad sólo una combinación de empaquetado de tuplas y desempaquetado de secuencias.

#### Diccionarios:

Otro tipo de dato útil es el diccionario. Los diccionarios se encuentran a vecer en otros lenguajes como *memorias asociativas* o *arreglos asociativos*. A diferencia de las secuencias, que se indexan mediante un rango numérico, **los diccionarios se indexan con claves**, que pueden ser cualquier tipo *inmutable*; las cadenas y número siempre pueden ser claves. Las tuplas pueden usarse como claves si solamente contienen cadenas, números o tuplas; **si una tupla contiene cualquier objeto mutable directa o indirectamente, no puede usarse como clave**. 

Es mejor pensar en un diccionario como un conjunto de *pares clave-valor* con el requerimiento de que las claves sean únicas (dentro de un diccionario). Un para de llaves crean un diccionario vacío `{}`. Colocar una lista de pares clave-valor separada por comas dentro de las llaves agrega, de inicio, pares clave-valor a nuestro diccionario; esta es también la forma en que los diccionarios se muestran en la salida.

**Las operaciones principales sobre un diccionario son guardar un valor con una clave y extraer ese valor dada la clave**. También es posible borrar un par clave-valor con `del`. **Si usas una clave que ya está en uso para guardar un valor, el valor que estaba asociado con esa clave se pierde.** Es un error extraer un valor usando una clave no existente.

Ejecutando `list(d)` en un diccionario retornará una lista con todas las claves usadas en el diccionario, en el orden de inserción. Para comprobar si una clave está en el diccionario se usa la palabra clave `in`.

In [83]:
dic1 = {'a': 40323, 'dasd': 21212}

In [84]:
dic1['qwerty'] = 12345

In [85]:
dic1

{'a': 40323, 'dasd': 21212, 'qwerty': 12345}

In [86]:
del dic1['dasd']

In [87]:
dic1['irv'] = 7890

In [88]:
dic1

{'a': 40323, 'qwerty': 12345, 'irv': 7890}

In [89]:
list(dic1)

['a', 'qwerty', 'irv']

In [90]:
sorted(dic1)

['a', 'irv', 'qwerty']

In [91]:
'a' in dic1

True

In [92]:
'dasd' in dic1

False

El constructor `dict()` crea un diccionario directamente desde secuencias de pares clave-valor:

In [93]:
dict( [ ('sape', 4139), ('guido', 12345), ('jack', 54321) ] )

{'sape': 4139, 'guido': 12345, 'jack': 54321}

Además, las comprensiones de diccionarios se pueden usar para crear diccionarios desde expresiones arbitrarias de clave-valor:

In [94]:
{x: x**2 for x in (2,4,6)}

{2: 4, 4: 16, 6: 36}

Cuando las claves son cadenas simples, a veces resulta más fácil especificar los pares usando argumentos por palabra clave:

In [95]:
dict(sape=12345, guido=54321, jack=6789)

{'sape': 12345, 'guido': 54321, 'jack': 6789}

---

Ahora sí, vamos a hablar sobre listas y diccionarios anidados, listas que contienen listas, listas que contienen diccionarios, diccionarios que almacenan lista, etc.

Hagamos un ejemplo de ello:

In [96]:
#Main
def run(): 
    my_list = [1, 'Holi', True, 4.5]
    my_dict = {"fristname": 'Abdiel', "Lastname": 'Guerr'}
    
    #Lista de diccionarios
    super_list = [
        {"fristname": 'Abdiel', "Lastname": 'Guerro'},
        {"fristname": 'ASDFG', "Lastname": 'gdfgdfg'},
        {"fristname": 'ghjkl', "Lastname": 'asdasdsa'},
        {"fristname": 'qwert', "Lastname": 'weqwe'},
        {"fristname": 'zxcvv', "Lastname": 'cvzczxc'},
        {"fristname": 'vbnmm', "Lastname": 'hfghfg'}, 
    ]

    #Diccionario de listas
    super_dict = {
        "natural_nums": [1,2,3,4,5],
        "integer_nums": [-1, -2, 0, 1, 2],
        "floating_nums": [1.1, 2.2, 3.3, 4.4]
    }
    
    # El método .items() nos permite recorrer llaves y valores
    # de un diccionario al mismo tiempo en un ciclo.
    print("Super_dict:")
    for key, value in super_dict.items():
        print(f'Key: {key}, Value: {value}')
        
        
    print("\nSuper_list:")
    #En cada iteración hay un diccionario
    indice = 0
    for elemento in super_list:
        indice = indice + 1
        print(f'\nDiccionario número: {indice} = {elemento}') 
        #Por cada diccionario tenemos un par llave-valor
        for key2, value2 in elemento.items(): 
            print(f'Key {indice}: {key2}, Value {indice}: {value2}')
   

# Entry point  
if __name__ == '__main__': 
    run()

Super_dict:
Key: natural_nums, Value: [1, 2, 3, 4, 5]
Key: integer_nums, Value: [-1, -2, 0, 1, 2]
Key: floating_nums, Value: [1.1, 2.2, 3.3, 4.4]

Super_list:

Diccionario número: 1 = {'fristname': 'Abdiel', 'Lastname': 'Guerro'}
Key 1: fristname, Value 1: Abdiel
Key 1: Lastname, Value 1: Guerro

Diccionario número: 2 = {'fristname': 'ASDFG', 'Lastname': 'gdfgdfg'}
Key 2: fristname, Value 2: ASDFG
Key 2: Lastname, Value 2: gdfgdfg

Diccionario número: 3 = {'fristname': 'ghjkl', 'Lastname': 'asdasdsa'}
Key 3: fristname, Value 3: ghjkl
Key 3: Lastname, Value 3: asdasdsa

Diccionario número: 4 = {'fristname': 'qwert', 'Lastname': 'weqwe'}
Key 4: fristname, Value 4: qwert
Key 4: Lastname, Value 4: weqwe

Diccionario número: 5 = {'fristname': 'zxcvv', 'Lastname': 'cvzczxc'}
Key 5: fristname, Value 5: zxcvv
Key 5: Lastname, Value 5: cvzczxc

Diccionario número: 6 = {'fristname': 'vbnmm', 'Lastname': 'hfghfg'}
Key 6: fristname, Value 6: vbnmm
Key 6: Lastname, Value 6: hfghfg


---

### List Comprehensions

Como practica intentemos crear una lista que contenga los primeros 100 números naturales al cuadrado.

In [97]:
def main():
    
    list_n_to_2 = []
    for i in range(1, 101):
        list_n_to_2.append(i**2)
    print(f'Lista de naturales al cuadrado:\n {list_n_to_2}')
    

if __name__ == '__main__':
    main()

Lista de naturales al cuadrado:
 [1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900, 961, 1024, 1089, 1156, 1225, 1296, 1369, 1444, 1521, 1600, 1681, 1764, 1849, 1936, 2025, 2116, 2209, 2304, 2401, 2500, 2601, 2704, 2809, 2916, 3025, 3136, 3249, 3364, 3481, 3600, 3721, 3844, 3969, 4096, 4225, 4356, 4489, 4624, 4761, 4900, 5041, 5184, 5329, 5476, 5625, 5776, 5929, 6084, 6241, 6400, 6561, 6724, 6889, 7056, 7225, 7396, 7569, 7744, 7921, 8100, 8281, 8464, 8649, 8836, 9025, 9216, 9409, 9604, 9801, 10000]


Ahora tratemos de hacer una lista con los cuadrados del 1 al 100 que NO sean divisibles entre 3.

In [98]:
def main():
    
    n_to_2_not_divisible3 = []
    for i in range (101):
        n_to_2_not_divisible3.append(i**2)
        if (i**2)%3 == 0:
            n_to_2_not_divisible3.remove(i**2)

    print('Lista de naturales al cuadrado que NO son divisibles entre 3\n' 
          , n_to_2_not_divisible3)
    
if __name__ == '__main__':
    main()

Lista de naturales al cuadrado que NO son divisibles entre 3
 [1, 4, 16, 25, 49, 64, 100, 121, 169, 196, 256, 289, 361, 400, 484, 529, 625, 676, 784, 841, 961, 1024, 1156, 1225, 1369, 1444, 1600, 1681, 1849, 1936, 2116, 2209, 2401, 2500, 2704, 2809, 3025, 3136, 3364, 3481, 3721, 3844, 4096, 4225, 4489, 4624, 4900, 5041, 5329, 5476, 5776, 5929, 6241, 6400, 6724, 6889, 7225, 7396, 7744, 7921, 8281, 8464, 8836, 9025, 9409, 9604, 10000]


Ahora, ¿Que pasaría si quisieramos realizar estas tareas de generar listas, pero **sin ciclos**?

In [99]:
def main():
    
    list_n_to_2 = [n**2 for n in range(101)]
    dict_n_to_2 = {n: n**2 for n in range(101)}
    
    print(f'Lista de naturales del 1-100 al cuadrado: \n{list_n_to_2}')
    print(f'\n Diccionario de naturales del 1-100 al cuadrado:\n{dict_n_to_2}')

if __name__ == '__main__':
    main()

Lista de naturales del 1-100 al cuadrado: 
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900, 961, 1024, 1089, 1156, 1225, 1296, 1369, 1444, 1521, 1600, 1681, 1764, 1849, 1936, 2025, 2116, 2209, 2304, 2401, 2500, 2601, 2704, 2809, 2916, 3025, 3136, 3249, 3364, 3481, 3600, 3721, 3844, 3969, 4096, 4225, 4356, 4489, 4624, 4761, 4900, 5041, 5184, 5329, 5476, 5625, 5776, 5929, 6084, 6241, 6400, 6561, 6724, 6889, 7056, 7225, 7396, 7569, 7744, 7921, 8100, 8281, 8464, 8649, 8836, 9025, 9216, 9409, 9604, 9801, 10000]

 Diccionario de naturales del 1-100 al cuadrado:
{0: 0, 1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100, 11: 121, 12: 144, 13: 169, 14: 196, 15: 225, 16: 256, 17: 289, 18: 324, 19: 361, 20: 400, 21: 441, 22: 484, 23: 529, 24: 576, 25: 625, 26: 676, 27: 729, 28: 784, 29: 841, 30: 900, 31: 961, 32: 1024, 33: 1089, 34: 1156, 35: 1225, 36: 1296, 37: 1369, 38: 1444, 39: 1521, 

Ahora hagamos el segundo ejemplo de los cuadrados que no son divisibles entre 3 usando list comprehensions:

In [100]:
def main():
    
    n_to_2_not_divisible3 = [n**2 for n in range(1,101) if n%3 != 0]
    print('Lista de naturales al cuadrado que NO son divisibles entre 3\n' 
          , n_to_2_not_divisible3)
    
if __name__ == '__main__':
    main()

Lista de naturales al cuadrado que NO son divisibles entre 3
 [1, 4, 16, 25, 49, 64, 100, 121, 169, 196, 256, 289, 361, 400, 484, 529, 625, 676, 784, 841, 961, 1024, 1156, 1225, 1369, 1444, 1600, 1681, 1849, 1936, 2116, 2209, 2401, 2500, 2704, 2809, 3025, 3136, 3364, 3481, 3721, 3844, 4096, 4225, 4489, 4624, 4900, 5041, 5329, 5476, 5776, 5929, 6241, 6400, 6724, 6889, 7225, 7396, 7744, 7921, 8281, 8464, 8836, 9025, 9409, 9604, 10000]


La estructura general de los list comprehensions sería la siguiente:


```py
[element for element in iterable if condition] 
> Para cada elemento en el iterable agrega ese elemento 
  solo si cumple la condición.
```

Hagamos otro ejemplo utilizando list comprehensions que enliste a todos los múltiplos de 4 que a la vez también son múltiplos de 6 y de 9, hasta 5 dígitos.

In [101]:
def main():
    list_muitiple_of_9_6_4 = [e for e in range(1,1000) if e%9 == 0 and 
                             e%6 == 0 and e%4 == 0]
    print(list_muitiple_of_9_6_4)
    

if __name__ == '__main__':
    main()

[36, 72, 108, 144, 180, 216, 252, 288, 324, 360, 396, 432, 468, 504, 540, 576, 612, 648, 684, 720, 756, 792, 828, 864, 900, 936, 972]


---

### Dictionary comprehensions

In [102]:
def main():
    dict_naturals = {key: key**3 for key in range(101)}
    print(f'Dictionary comprehensions:\n{dict_naturals}')
    
if __name__ == '__main__':
    main()

Dictionary comprehensions:
{0: 0, 1: 1, 2: 8, 3: 27, 4: 64, 5: 125, 6: 216, 7: 343, 8: 512, 9: 729, 10: 1000, 11: 1331, 12: 1728, 13: 2197, 14: 2744, 15: 3375, 16: 4096, 17: 4913, 18: 5832, 19: 6859, 20: 8000, 21: 9261, 22: 10648, 23: 12167, 24: 13824, 25: 15625, 26: 17576, 27: 19683, 28: 21952, 29: 24389, 30: 27000, 31: 29791, 32: 32768, 33: 35937, 34: 39304, 35: 42875, 36: 46656, 37: 50653, 38: 54872, 39: 59319, 40: 64000, 41: 68921, 42: 74088, 43: 79507, 44: 85184, 45: 91125, 46: 97336, 47: 103823, 48: 110592, 49: 117649, 50: 125000, 51: 132651, 52: 140608, 53: 148877, 54: 157464, 55: 166375, 56: 175616, 57: 185193, 58: 195112, 59: 205379, 60: 216000, 61: 226981, 62: 238328, 63: 250047, 64: 262144, 65: 274625, 66: 287496, 67: 300763, 68: 314432, 69: 328509, 70: 343000, 71: 357911, 72: 373248, 73: 389017, 74: 405224, 75: 421875, 76: 438976, 77: 456533, 78: 474552, 79: 493039, 80: 512000, 81: 531441, 82: 551368, 83: 571787, 84: 592704, 85: 614125, 86: 636056, 87: 658503, 88: 681472, 8

Igual que como hicimos con las listas, tratemos de agregar en un diccionario aquellos números al cubo que no sean divisibles entre 3.

In [103]:
def main():
    dict_not_divisible_3 = {n:n**3 for n in range(0,101) if n%3 != 0}
    print(dict_not_divisible_3)
    
    
if __name__ == '__main__':
    main()

{1: 1, 2: 8, 4: 64, 5: 125, 7: 343, 8: 512, 10: 1000, 11: 1331, 13: 2197, 14: 2744, 16: 4096, 17: 4913, 19: 6859, 20: 8000, 22: 10648, 23: 12167, 25: 15625, 26: 17576, 28: 21952, 29: 24389, 31: 29791, 32: 32768, 34: 39304, 35: 42875, 37: 50653, 38: 54872, 40: 64000, 41: 68921, 43: 79507, 44: 85184, 46: 97336, 47: 103823, 49: 117649, 50: 125000, 52: 140608, 53: 148877, 55: 166375, 56: 175616, 58: 195112, 59: 205379, 61: 226981, 62: 238328, 64: 262144, 65: 274625, 67: 300763, 68: 314432, 70: 343000, 71: 357911, 73: 389017, 74: 405224, 76: 438976, 77: 456533, 79: 493039, 80: 512000, 82: 551368, 83: 571787, 85: 614125, 86: 636056, 88: 681472, 89: 704969, 91: 753571, 92: 778688, 94: 830584, 95: 857375, 97: 912673, 98: 941192, 100: 1000000}


Podemos inferir que la estructura de los *Dictionary comprehensions* es la siguiente:

```py
{key : value for value in iterable if condition}
>  Para cada elemento en un iterable, coloca una llave y un valor 
   solo si cumple la condición.
```

Como ejercicio realicemos lo siguiente:

* Crear un *Dictionary comprehension* que contenga en sus llaves a los primeros 1000 números naturales con sus raíces cuadradas como valores.

In [104]:
import math 
from math import sqrt

def main():
    dict_naturals_squareroots = {n: sqrt(n) for n in range(11) }
    #Lo truncamos a 10 para que no ocupe mucho espacio al imprimir
    print('Diccionario con los naturales en sus llaves y la raiz en los valores:\n\n'
            ,dict_naturals_squareroots)


if __name__ == '__main__':
    main()

Diccionario con los naturales en sus llaves y la raiz en los valores:

 {0: 0.0, 1: 1.0, 2: 1.4142135623730951, 3: 1.7320508075688772, 4: 2.0, 5: 2.23606797749979, 6: 2.449489742783178, 7: 2.6457513110645907, 8: 2.8284271247461903, 9: 3.0, 10: 3.1622776601683795}


---

## Módulo 4. Conceptos avanzados de funciones

### Funciones anónimas: lambda

Recordemos un poco el concepto de función. Una función es un bloque de código con un nombre asociado, que recibe creo o más argumentos como entrada, sigue una secuencia de sentencias, las cuales ejecutan una operación deseada y devuelve un valor y/o realiza una tarea. 

En otras palabras, es un fragmento de código que escribimos una sola vez y puede ser reutilizada en el resto de nuestro programa, donde podemos cambiar nuestros parámetros de entrada para obtener diferentes salidas.

La estructura básica de una función es la siguiente:

```py
def function_name(parameter_1, parameter_2, ... , parameter_n)

> Define una función (keyword def) con cierto nombre (identificador)
  y ciertos parámetros
```

Puede darse el caso donde utilicemos funciones que no reciben ningún parámetro de entrada.

Pero existe una forma diferente para crear funciones, más específicamente, **funciones sin nombre**. A estas funciones les llamamos **funciones anónimas o lambda functions**.

Es decir, solo son funciones que carecen de un identificador. Además de tener ciertas particularidades.

La estructura de las funciones lambda $\lambda$ es la siguiente:

```py
lambda argumentos : expresión 
```

En Python las funciones $\lambda$ necesariamente deben de ser escritas en una sola línea de código.

Para hacer notar la diferencia, recordemos un pequeño programa donde verificamos si una palabra era palíndromo o no.

Primero revisemos el programa hecho con funciones normales, teniendo la función `palindromo(palabra)` que hace la verificación y la función `main()` que ejecuta a `palindromo()` y crea los intputs en donde pasamos la palabra a verificar.

En realidad podría funcionar sin la función `main()` agregando los inputs en el entry point `if __name__ == '__main__'` , pero como buena práctica y por legibilidad lo dejamos así.

In [106]:
def palindromo(palabra):
    palabra = palabra.replace(' ','') #Reemplaza los espacios por un vacio
    palabra = palabra.lower() #Ejecutamos el metodo de minusculas
    palabra_invertida = palabra[::-1] #Guardar la palabra al reves, de inicio a final con pasos de -1

    if palabra == palabra_invertida:
        return True
    else:
        return False


def main():
    palabra = input('Escribe una palabra: ')
    es_palindromo = palindromo(palabra)

    if es_palindromo == True:
        print('Es un palindromo')
    
    else:
        print('No es un palindromo')


if __name__ == '__main__':
    main()

Escribe una palabra: ala
Es un palindromo


Ahora, revisemos como sería la implementación con funciones lambda:

```Python
palindrome = lambda string: string = string[::-1]
```

Donde tenemos las siguientes partes:
   
   * **Identificador:** Nombre *asociado a la variable que contiene a un objeto del tipo función que retorna el resultado* no a la función en sí (`palindrome `).
   
   
   * **Argumento:** Parámetro que recibe la función para trabajar, acompañado del keyword `lambda` (`lambda string`).
   
       
   * **Expresión:** Línea de código que indica como va a trabajar (`string = string[::-1]`).

In [107]:
palindrome = lambda string : string == string[::-1]
print(palindrome('larutanatural'))

True


---

### High order functions: filter, map y reduce

Una función de orden superior es una función que recibe como parámetro a otra función.

Existen 3 tipos de funciones de orden superior que son muy importantes en una gran parte de los lenguajes de programación:

#### `filter()`:

Es una función la cual toma un *predicado y una lista* y devuelve una lista con los elementos que satisfacen el predicado. Tal como su nombre lo indica `filter()` significa "filtrar", ya que a partir de una lista o iterador y una función condicional, es capaz de devolver una nueva colección de elementos filtrados que cumplan la condición.

Todo esto podría haberse logrado también usando **list comprehension** que usaran predicados. No hay ninguna regla de cuando usar una u otra opción, simplemente se debe elegir que es más legible y práctico para la situación.

Como ejemplo imaginemos que queremos filtrar de los datos de una lista aquellos números pares y solo dejar a los números impares.

![](https://i.imgur.com/sn1MOU8.png)

Primero una solución con list comprehensions:

In [108]:
my_list = [n for n in range(11)]

add = [i for i in my_list if i % 2 != 0]

print(add)


[1, 3, 5, 7, 9]


Ahora, resolvamos usando la función de orden superior `filter()`:

In [109]:
my_list = [n for n in range(11)]

add = list(filter(lambda x: x%2 !=0, my_list))

print(add)

[1, 3, 5, 7, 9]


Podemos ver que dentro de nuestra función `filter()` está recibiendo como primer parámetro a una función lambda $\lambda$, la cual recibe como parámetro a una `x` y retorna como resultado a `x%2 != 0`; como segundo parámetro de `filter()` está recibiendo a una lista (iterable) que no es más que nuestra lista de números del 1 al 10.

La función `filter()` devuelve como resultado a un objeto del tipo **iterador**. Y toda nuestra función `filter()` está dentro de otra función llamada `list()` que se encarga de convertir nuestro iterable en una lista para finalmente almacenar todo el resultado dentro de nuestra variable `add`.

#### `map()`:

La función `map()` toma una función y una lista y aplica esa función a cada elemento de esa lista, produciendo una nueva lista como resultado.

Funciona de una forma muy similar a `filter()`, con la diferencia de que en lugar de aplicar una condición a un elemento de una lista o secuencia, aplica una función sobre todos los elementos y como resultado devuelve una lista de los resultados de efectuar esa función.

Imaginemos que originalmente tenemos una lista con ciertos números y queremos obtener una nueva lista, pero ahora con los números elevados al cuadrado.

![](https://i.imgur.com/QgD5vJa.png)

Veamos como lo resolveríamos con *list comprehensions*:

In [110]:
def main():
    numbers = [n for n in range(11)]
    print(f'Lista de números:\n{numbers}')
    
    numbers_2 = [x**2 for x in numbers]
    print(f'Lista de números al cuadrado:\n{numbers_2}')


if __name__ == '__main__':
    main()

Lista de números:
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Lista de números al cuadrado:
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


Ahora veamos como resolverlo con `map()`:

In [111]:
def main():
    numbers = [n for n in range(11)]
    print(f'Lista de números:\n{numbers}')
    
    numbers_2 = list(map(lambda x: x**2, numbers))
    print(f'Lista de números al caudrado:\n{numbers_2}')

    
if __name__ == '__main__':
    main()

Lista de números:
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Lista de números al caudrado:
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


#### `reduce()`:

Esta función toma como argumento un conjunto de valores (una lista, tupla o cualquier objeto iterable) y lo "reduce" a un único valor. El cómo se obtiene ese único valor a partir de la colección pasada como argumento dependerá de la función aplicada (Segundo parámetro).

Imaginemos el ejemplo donde tenemos una lista de números y queremos obtener el resultado de multiplicar cada uno de ellos, es decir, el primero por el segundo por el tercero, etc.

![](https://i.imgur.com/BZslNRg.png)

Veamos como sería una implementación común:

In [112]:
def main():
    numbers = [n for n in range(1,11)]
    print(f'Numeros: {numbers}')
    
    multiply_all = 1
    
    for i in numbers:
        multiply_all = multiply_all * i
    
    print(f'Resultado: {multiply_all}')


if __name__ == '__main__':
    main()

Numeros: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Resultado: 3628800


Ahora veamos la implementación usando la función `reduce()`. En este caso es un tanto diferente ya que tenemos que importar la función del módulo `functools` antes de usarla.

Recibirá a la función lambda como primer parámetro y a nuestra lista de números como segundo parámetro.

Lambda recibe dos parámetros: `a` y `b`, es decir, el primer y segundo elemento de la lista. En este caso en particular comenzamos con `1*2=2`, después con`2*3=6`, luego con `6*4=24` y así sucesivamente.

In [113]:
from functools import reduce

def main():
    numbers = [n for n in range(1,11)]
    print(f'Números: {numbers}')
    
    multiply_all = reduce(lambda a, b: a*b , numbers)
    print(f'Resultado:{multiply_all}')


if __name__ == '__main__':
    main()

Números: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Resultado:3628800


---

### Proyecto: Filtrando datos

Para comenzar este pequeño proyecto vamos a trabajar con esta lista de diccionarios de datos para comenzar a filtrarlos:

In [114]:
DATA = [
    {
        'name': 'Facundo',
        'age': 72,
        'organization': 'Platzi',
        'position': 'Technical Coach',
        'language': 'python',
    },
    {
        'name': 'Luisana',
        'age': 33,
        'organization': 'Globant',
        'position': 'UX Designer',
        'language': 'javascript',
    },
    {
        'name': 'Héctor',
        'age': 19,
        'organization': 'Platzi',
        'position': 'Associate',
        'language': 'ruby',
    },
    {
        'name': 'Gabriel',
        'age': 20,
        'organization': 'Platzi',
        'position': 'Associate',
        'language': 'javascript',
    },
    {
        'name': 'Isabella',
        'age': 30,
        'organization': 'Platzi',
        'position': 'QA Manager',
        'language': 'java',
    },
    {
        'name': 'Karo',
        'age': 23,
        'organization': 'Everis',
        'position': 'Backend Developer',
        'language': 'python',
    },
    {
        'name': 'Ariel',
        'age': 32,
        'organization': 'Rappi',
        'position': 'Support',
        'language': '',
    },
    {
        'name': 'Juan',
        'age': 17,
        'organization': '',
        'position': 'Student',
        'language': 'go',
    },
    {
        'name': 'Pablo',
        'age': 32,
        'organization': 'Master',
        'position': 'Human Resources Manager',
        'language': 'python',
    },
    {
        'name': 'Lorena',
        'age': 56,
        'organization': 'Python Organization',
        'position': 'Language Maker',
        'language': 'python',
    },
]

Ahora sí, comencemos a crear el programa

In [115]:
DATA = [
    {
        'name': 'Facundo',
        'age': 72,
        'organization': 'Platzi',
        'position': 'Technical Coach',
        'language': 'python',
    },
    {
        'name': 'Luisana',
        'age': 33,
        'organization': 'Globant',
        'position': 'UX Designer',
        'language': 'javascript',
    },
    {
        'name': 'Héctor',
        'age': 19,
        'organization': 'Platzi',
        'position': 'Associate',
        'language': 'ruby',
    },
    {
        'name': 'Gabriel',
        'age': 20,
        'organization': 'Platzi',
        'position': 'Associate',
        'language': 'javascript',
    },
    {
        'name': 'Isabella',
        'age': 30,
        'organization': 'Platzi',
        'position': 'QA Manager',
        'language': 'java',
    },
    {
        'name': 'Karo',
        'age': 23,
        'organization': 'Everis',
        'position': 'Backend Developer',
        'language': 'python',
    },
    {
        'name': 'Ariel',
        'age': 32,
        'organization': 'Rappi',
        'position': 'Support',
        'language': '',
    },
    {
        'name': 'Juan',
        'age': 17,
        'organization': '',
        'position': 'Student',
        'language': 'go',
    },
    {
        'name': 'Pablo',
        'age': 32,
        'organization': 'Master',
        'position': 'Human Resources Manager',
        'language': 'python',
    },
    {
        'name': 'Lorena',
        'age': 56,
        'organization': 'Python Organization',
        'position': 'Language Maker',
        'language': 'python',
    },
]


def run():
    
    #Worker = datos de cada trabajador individual
    all_python_devs = [worker["name"] for worker in DATA
                      if worker["language"] == "python"] 
    
    all_platzi_workers = [worker["name"] for worker in DATA
                         if worker["organization"] == "Platzi"]
    
    adults = list(filter(lambda worker: worker["age"] > 18 , DATA))
    adults_name = list(map(lambda worker: worker["name"], adults))
    
    # El simbolo | sirve para unir diccionarios
    # Solo funciona con Python >= 3.9
    #old_people = list(map(lambda worker: worker | {"old": worker["age"] > 70}, DATA))
    old_people = list(map(lambda worker:
                          {**worker, **{"old": worker["age"] > 70}}, DATA))
    
    print(f'All Python Devs: {all_python_devs}')
    print(f'All Platzi Workers: {all_platzi_workers}')
    print(f'Adults: {adults_name}')
    #print(f'Olde People: {old_people}')



if __name__ == '__main__':
    run()

All Python Devs: ['Facundo', 'Karo', 'Pablo', 'Lorena']
All Platzi Workers: ['Facundo', 'Héctor', 'Gabriel', 'Isabella']
Adults: ['Facundo', 'Luisana', 'Héctor', 'Gabriel', 'Isabella', 'Karo', 'Ariel', 'Pablo', 'Lorena']


Ahora hagamos el mismo ejercicio un poco a la inversa, hagamos las listas `all_python_devs` y `all_platzi_workers` usando las funciones de orden superior `filer()` y `map()`.

Y para el caso de `old_people` y `adults` hagamoslas con list comprehensions.

In [116]:
DATA = [
    {
        'name': 'Facundo',
        'age': 72,
        'organization': 'Platzi',
        'position': 'Technical Coach',
        'language': 'python',
    },
    {
        'name': 'Luisana',
        'age': 33,
        'organization': 'Globant',
        'position': 'UX Designer',
        'language': 'javascript',
    },
    {
        'name': 'Héctor',
        'age': 19,
        'organization': 'Platzi',
        'position': 'Associate',
        'language': 'ruby',
    },
    {
        'name': 'Gabriel',
        'age': 20,
        'organization': 'Platzi',
        'position': 'Associate',
        'language': 'javascript',
    },
    {
        'name': 'Isabella',
        'age': 30,
        'organization': 'Platzi',
        'position': 'QA Manager',
        'language': 'java',
    },
    {
        'name': 'Karo',
        'age': 23,
        'organization': 'Everis',
        'position': 'Backend Developer',
        'language': 'python',
    },
    {
        'name': 'Ariel',
        'age': 32,
        'organization': 'Rappi',
        'position': 'Support',
        'language': '',
    },
    {
        'name': 'Juan',
        'age': 17,
        'organization': '',
        'position': 'Student',
        'language': 'go',
    },
    {
        'name': 'Pablo',
        'age': 32,
        'organization': 'Master',
        'position': 'Human Resources Manager',
        'language': 'python',
    },
    {
        'name': 'Lorena',
        'age': 56,
        'organization': 'Python Organization',
        'position': 'Language Maker',
        'language': 'python',
    },
]

def run():
    all_python_devs = list(filter(lambda worker: worker["language"]=="python", DATA))
    all_python_devs_names = list(map(lambda worker: worker["name"], all_python_devs))
    
    all_platzi_workers = list(filter(lambda worker:
                                     worker["organization"] == "Platzi" , DATA))
    all_platzi_workers_name = list(map(lambda worker:
                                      worker["name"] , all_platzi_workers))
    adult = [worker["name"] for worker in DATA if worker["age"]>18]
    
    old_people = [{**worker, **{'old': worker['age'] > 70}} for worker in DATA]
    
    print(f'All Python Devs: {all_python_devs_names}')
    print(f'All Platzi workers: {all_platzi_workers_name}')
    print(f'Adult People: {adult}')
    print(f'\nNew Key: Old People\n {old_people}')

  
if __name__ == '__main__':
    run()

All Python Devs: ['Facundo', 'Karo', 'Pablo', 'Lorena']
All Platzi workers: ['Facundo', 'Héctor', 'Gabriel', 'Isabella']
Adult People: ['Facundo', 'Luisana', 'Héctor', 'Gabriel', 'Isabella', 'Karo', 'Ariel', 'Pablo', 'Lorena']

New Key: Old People
 [{'name': 'Facundo', 'age': 72, 'organization': 'Platzi', 'position': 'Technical Coach', 'language': 'python', 'old': True}, {'name': 'Luisana', 'age': 33, 'organization': 'Globant', 'position': 'UX Designer', 'language': 'javascript', 'old': False}, {'name': 'Héctor', 'age': 19, 'organization': 'Platzi', 'position': 'Associate', 'language': 'ruby', 'old': False}, {'name': 'Gabriel', 'age': 20, 'organization': 'Platzi', 'position': 'Associate', 'language': 'javascript', 'old': False}, {'name': 'Isabella', 'age': 30, 'organization': 'Platzi', 'position': 'QA Manager', 'language': 'java', 'old': False}, {'name': 'Karo', 'age': 23, 'organization': 'Everis', 'position': 'Backend Developer', 'language': 'python', 'old': False}, {'name': 'Ariel', 

---

## Módulo 5. Manejo de errores

### Los errores en el código

Vamos a dividir los posibles errores en 2 categorías:

* **Python nos avisa que hay un error:** Python nos devuelve un traceback que contiene más información sobre el error.

* **Python NO nos avisa sobre el error:** Es decir, hay un *error de lógica* dentro del funcionamiento de nuestro algoritmo.

Para el primer caso, donde recibimos una notificación de Python; vamos a dividir los problemas de la siguiente manera:

![](https://i.imgur.com/Mj0x0g9.png)

Las ramas principales de nuestros errores son los **errores de sintaxis** (está mal escrita alguna parte del código) y las **excepciones**.

Existe una diferencia principal entre estos 2 tipos de errores. Cuando se trata de un error de sintaxis **el programa se detiene antes de que Python ejecute la primer línea de código**, mientras que en un error de excepciones hace que la lógica del programa de quiebre **una vez llegando a esa linea de código en particular**, por ejemplo, cuando dividimos entre 0.

Vamos a basarnos en la [documentación de Python](https://docs.python.org/es/3/tutorial/errors.html) para revisar lo siguiente:

#### Errores de sintaxis

Los errores de sintaxis, también conocidos como errores de interpretación, son quizá el tipo de queja más común que tendremos.

Veamos un ejemplo

In [117]:
while True print("Holi")

SyntaxError: invalid syntax (<ipython-input-117-ee70e2354ab3>, line 1)

Lo que pasa aquí es que el intérprete regresa la línea responsable del error (la parte amarilla) y muestra una pequeña flecha que apunta al primer lugar donde se detectó el error (punta de flecha gris). En este caso el error ha sido provocado (o al menos detectado) en el elemento que precede a la flecha que es la instrucción `print` ya que nos hicieron falta los dos puntos `:` antes del mismo. La instrucción correcta debió ser `while True: print("Holi")`.

Además se nos muestra el nombre del archivo ` File "<ipython-input-6-ee70e2354ab3>"` (en este caso una celda del notebook) y la línea donde se detectó el error `line 1`.

#### Excepciones

Incluso si una declaración o expresión es sintácticamente correcta, puede generar un error cuando se intenta ejecutar. **Los errores detectados durante la ejecución** se llaman *excepciones*, y no son incondicionalmente fatales. Sin embargo la mayoría de excepciones no son gestionadas por el código, y resultan en mensajes de error como los siguientes:

In [118]:
10 * (1/0)

ZeroDivisionError: division by zero

In [119]:
4 + spam*3

NameError: name 'spam' is not defined

In [120]:
'2' + 2

TypeError: can only concatenate str (not "int") to str

La última línea de los mensajes de error indica qué ha sucedido. Hay excepciones de diferentes tipos, y el tipo se imprime como parte del mensaje: `ZeroDivisionError`, `NameError` y `TypeError`. La cadena mostrada como tipo de la excepción es el nombre de la excepción predefinida que ha ocurrido. Esto es válido para todas las excepciones predefinidas del intérprete, pero no tiene por que ser así para excepciones definidas por el usuario.

El resto de la línea provee información basado en el tipo de la excepción que la causó.

---

Vamos a ver las principales excepciones que nos podemos encontrar en Python:

* `KeyboardInterrupt`: Esta aparece cuando nosotros pulsamos `CTRL + C` y detenemos la consola interactiva de Python. Es decir, cortamos el proceso y cerramos todo, esto lo logra **elevando** una excepción.

    A lo que nos referimos con elevar es que Python internamente crea un objeto de tipo *excepcion* o error y lo va moviendo a través de nuestro programa desde los bloques de dentro hacía afuera (le va a ir dando más peso).



* `KeyError`: Este error ocurre cuando tratamos de acceder a una llave que no existe dentro de un diccionario.



* `IndexError`: Esta es el caso anterior, pero en listas. Ocurre cuando tratamos de acceder a un índice que no existe dentro de una lista.


* `FileNotFoundError`: Ocurre cuando tratamos de abrir un archivo en Python que no existe dentro de la ruta especificada.



* `ZeroDivisionError`: Error que aparece cuando tratamos de realizar una división entre 0 (matemáticamente indefinida).



* `ImportError`: Ocurre cuando intentamos importar un módulo y éste está mal escrito, tiene errores con el resto de módulos, etc.


para revisar el resto de excepciones [revisar la documentación](https://docs.python.org/es/3/library/exceptions.html#bltin-exceptions).

#### Traceback

Este es el nombre que recibe el mensaje que se arroja en pantalla cuando ocurre un error. Y generalmente tiene la siguiente forma aunque a veces puede ser un mensaje mucho más largo

![](https://i.imgur.com/YXneQjo.png)

La forma correcta de leer el traceback es leyéndolo desde el final hacía el inicio. Ya que hasta el final de la línea se nos índica el tipo de excepción que ha ocurrido, antes de ella tenemos el archivo, la línea y el módulo en los cuales ocurrió el error. 

---

### Debugging o depuración

Ahora, vamos a concentrarnos en como resolver aquellos errores de lógica en donde Python no nos avisa que tenemos un error ya que las instrucciones están bien escritas, pero estamos obteniendo resultados indeseables. Para ello haremos uso de algo conocido como debugging.

Para entender este concepto vamos a crear un programa que esté mal hecho adrede y aprovecharemos una herramienta incorporada dentro de VisualStudioCode para hacer debugging.

Nuestro programa tratará de recibir un número como parámetro y retornar una lista con todos los posibles divisores de ese número:

In [122]:
def divisors(num):
    divisors = []
    for i in range (1, num + 1):
        if num % i == 1:
            divisors.append(i)
    return divisors

def run():
    num = int(input('Ingrese un número para obtener sus divisores: '))
    print(divisors(num))


if __name__ == '__main__':
    run()

Ingrese un número para obtener sus divisores: 10
[3, 9]


Para arreglar el problema usaremos una herramienta de VS para hacer debugging que tiene el siguiente icono en la parte lateral izquierda:

![](https://i.imgur.com/FJGRVbZ.png)

Nos abrirá el siguiente recuadro y daremos click en el botón azul "run and Debug"

![](https://i.imgur.com/whWLm4q.png)

Seleccionamos la opción donde dice `Python file` 

![](https://i.imgur.com/Wwd6Jpq.png)

Y vamos a ver que tenemos las siguientes opciones además de que el programa comienza a correr en la terminal de VS para debugging:

![](https://i.imgur.com/Z6UeDBI.png)

![](https://i.imgur.com/WAscUa1.png)

Si presionamos el botón de pausa y luego ingresamos nuestro número a evaluar veremos lo siguiente:

![](https://i.imgur.com/KOOb5oW.png)

la linea `print(divisor(num))` se marca de amarillo. Esto nos índica que el programa se ejecutó, pero está parado en esa línea. 

Además podemos ver que la barra superior habilita el resto de sus opciones:

![](https://i.imgur.com/SkTpgBt.png)

Como en este momento estamos en la instrucción `print(divisor(num))` y sabemos que ahí está ocurriendo nuestro error, presionamos la flecha hacía abajo (Step into o "Depurar paso a paso por instrucciones"), también funciona presionando F11.

Esto nos llevará a la función a la que hace referencia nuestro print anterior, de forma que ahora estamos dentro de la función `divisors()` y ahora podemos visualizar en la parte izquierda el como se van comportando nuestras variables a lo largo del programa.

![](https://i.imgur.com/i8RYsge.png)

Ahora, presionando el botón de "depurar paso a paso por procedimientos" o con F10. Podremos ver como se comporta esa sección del programa paso a paso y evaluar en que punto arroja resultados diferentes a los que nosotros esperamos.

Al hacer el debugging nos damos cuenta que la instrucción `if num % i == 1:` está mal, y que la línea correcta debería de ser `if num % i == 0:` ya que la operación `%` índica los restantes de una división, por lo que al tener un número que sea divisor éste tendrá que ser = 0.

Entonces nuestro programa quedaría de la siguiente forma:

In [42]:
def divisors(num):
    divisors = []
    for i in range (1, num + 1):
        if num % i == 0:
            divisors.append(i)
    return divisors

def run():
    num = int(input('Ingrese un número para obtener sus divisores: '))
    print(divisors(num))


if __name__ == '__main__':
    run()

Ingrese un número para obtener sus divisores: 10
[1, 2, 5, 10]


Y ahora sí está funcionando correctamente.

Como complemento adicional revisemos otra opción que nos permite el debugger de VS. Hablemos de los **break points**.

#### BreakPoints

Imaginemos que tenemos un programa de 1000 lineas de código y a nosotros nos interesa revisar que pasa en la línea número 630, pero no queremos recorrer todo el programa hasta llegar ahí. Entonces, ¿Qué pasaría si quiero pausar en una línea específica de mi programa al hacer el debuggin? 

Para ello, **antes de empezar el debugging** vamos a presionar en la parte izquierda de nuestras líneas de código el círculito rojo en  la línea hasta la que queremos llegar. Éste círculo rojo se llama *punto de interrupción*.

![](https://i.imgur.com/MMZRGC2.png)

En este caso particular, estamos colocando nuestro **breakpoint** en la instrucción `if num % i == 0:` dentro de nuestro *for loop*.

Una vez colocado el breakpoint corremos nuestro debugger como lo hicimos anteriormente y al ejecutar nuestro programa veremos como se detiene en esa línea que marcamos, que en este caso es la línea 4. Esta herramienta nos ayudará a agilizar nuestro proceso de depuración.

---

### Manejo de excepciones

Para ello comencemos hablando de las palabras clave `try` y `except`.

#### `Try / Except`:

La declaración `try-except` funciona de la siguiente forma:

* Primero, se ejecuta la cláusula `try`.
* Si no ocurre ninguna excepción, la cláusula `except` se omite y la ejecución de la cláusula `try` finaliza.
* Si ocurre una excepción durante la ejecución de la cláusula `try`, el resto de la cláusula será omitida. Después, si el tipo de excepción coincide con las excepción indicada después de la cláusula `except`,  ejecutamos a `except` después de `try`.
* Si ocurre una excepción que no coincide con la indicada en la cláusula `except` se pasa a los `try` más externos; si no se encuentra un gestor, se genera una *unhandled  exception* (excepción no gestionada) y la ejecución se interrumpe con un mensaje.
* Una declaración `try` puede tener más de un `except`, para especificar gestores para distintas excepciones. A lo sumo un bloque será ejecutado. Sólo se gestionarán excepciones que ocurren en el correspondiente `try`, no en otros bloques del mismo `try`. Un `except` puede nombrar múltiples excepciones usando paréntesis, por ejemplo:

```py
except (RuntimeError, TypeError, NameError):
        pass
```

Revisemos un caso en donde Python nos arrojaría una excepción de este estilo en pantalla:

In [31]:
def palindrome(string):
    return string == string[::-1]

print(palindrome(1))

TypeError: 'int' object is not subscriptable

En este caso la función espera recibir un dato de tipo `string` en la entrada, no un tipo `int`. Este error podría ocurrir al crear una aplicación o programa para un cliente y que éste ingrese algún dato que nosotros nos esperabamos que ingresara, por lo que tenemos que anticiparnos a ello.

Para ello utilizaremos las keywords `try` y `except`.

In [32]:
def palindrome(string):
    return string == string[::-1]

try:
    print(palindrome(1)) 
except TypeError:
    print('Solo se pueden ingresar strings')

Solo se pueden ingresar strings


También podemos agregar un `else` o un `finally` a nuestro `try-except` de forma que hará lo siguiente:

* `try`: Aquí colocamos el código que esperamos que pueda lanzar un error.


* `except`: Aquí anticipamos y manejamos el posible error, es decir, si llega a ocurrir un error dentro del bloque `try`, se deja de ejecutar el código y ejecutamos lo que tengamos definido en `except`.


* `else`: Esta parte se ejecutará solo sí no hubo ninguna excepción lanzada desde el `try`.


* `finally`: Se ejecuta SIEMPRE, haya sido lanzada o no la excepción.

In [33]:
def palindrome(string):
    return string == string[::-1]

try:
    print(palindrome(1)) 
    #print(palindrome('1')) 
except TypeError:
    print('Solo se pueden ingresar strings')
else: 
    print('No ocurrió ninguna excepción :)')
finally:
    print('Esta línea se ejecuta SIEMPRE')

Solo se pueden ingresar strings
Esta línea se ejecuta SIEMPRE


#### `raise`:

La declaración `raise()` permite al programador a forzar la ocurrencia de una excepción específica, por ejemplo:

```py
>>> raise NameError('HiThere')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: HiThere
```

El único argumento a plantear indica la excepción que vamos a forzar. Éste debe de ser una *instancia de excepción o una clase de excepción*. Si se pasa una clase de excepción, se instanciará implícitamente llamando a su constructor sin argumentos:

```py
raise ValueError  

```

Ahora veamos que pasa cuando tenemos un error en nuestro programa, pero Python no lo detecta en sí como un error y queremos anticiparlo.

In [34]:
def palindrome(string):
    return string == string[::-1]

try:
    print(palindrome("")) 
except TypeError:
    print('Solo se pueden ingresar strings')

True


En este caso si estamos pasando una cadena de caracteres, pero es una cadena vacía. 

Al final del día Python si está realizando correctamente la tarea, ya que una cadena vacía es igual que una cadena vacía al revés, por ello nos devuelve true, pero nosotros sabemos que no es correcto que el usuario simplemente presione enter y deje el input vacío.

Para ello incorporaremos la keyword `raise`.

In [35]:
def palindrome(string):
    
    try:
        if len(string) == 0:
            raise ValueError("No se puede ingresar una cadena vacía")
        return string == string[::-1]
    
    except ValueError as ve:
        print(ve)
        return False

try: 
    print(palindrome(""))
except:
    print("Solo se pueden ingresar strings")

No se puede ingresar una cadena vacía
False


Podemos ver que ahora tenemos un nuevo `try-except` que engloba a la función de arriba.

En el bloque `try` decimos, "verificar si la longitud del string es igual a 0 entonces, **eleva** un error del tipo `ValueError` y en el paréntesis agrega el mensaje a mostrar junto con el error ("No se puede ingresar una cadena vacía") ;  si la longitud del string es diferentes de 0, regresa el resultado de la expresión `string == string[::-1]`".

Ahora, en el bloque`except` decimos, excepto de que haya ocurrido un `ValueError` como un `ve` es decir, estamos nombrando a nuestro error anterior de tipo `ValueError` con el nombre `ve`; imprime (`print`) el mensaje correspondiente a ese error y retorna `False` ya que para los fines del programa un espacio en blanco no corresponde con un palíndromo.

#### `finally`:

Esta keyword es muy rara de utilizar y se utiliza generalmente, al final de una estructura `try-except` que se puede utilizar para cosas como:

* Cerrar un archivo dentro de Python.
* Cerrar una conexión a una base de datos.
* Liberar recursos externos.

Y está instrucción dentro de `finally` **siempre se ejecutará al final de nuestra estructura `try-except`** suceda o no un `except`.

Veamos un ejemplo:

```py
try:
    f = open("archivo.txt")
finally:
    f.close()
```

### Poniendo a prueba el manejo de excepciones

¿Cómo nos anticiparíamos en el siguiente programa a que el usuario ingrese valores que no sean números, por ejemplo una letra o string?

In [38]:
def divisors(num):
    divisors = []
    for i in range (1, num + 1):
        if num % i == 0:
            divisors.append(i)
    return divisors

def run():
    num = int(input('Ingrese un número para obtener sus divisores: '))
    print(divisors(num))


if __name__ == '__main__':
    run()

Ingrese un número para obtener sus divisores: h


ValueError: invalid literal for int() with base 10: 'h'

Podemos ver que tenemos lo siguiente:

```Python
ValueError: invalid literal for int() with base 10: 'h'
```

Es decir, tenemos un error del tipo `ValueError`, lo cual es de esperarse ya que estamos esperando a recibir un valor numérico y le hemos ingresado una letra.

¿Cómo lo arreglamos? Bueno, empecemos por agregar toda nuestra función en un bloque `try-except` para trabajar con las excepciones. 

En este caso nuestra excepción es un `ValueError`.

In [39]:
def divisors(num):
    divisors = []
    for i in range (1, num + 1):
        if num % i == 0:
            divisors.append(i)
    return divisors

def run():
    try: 
        num = int(input('Ingrese un número para obtener sus divisores: '))
        print(divisors(num))
        
    except ValueError:
        print("\nʕ•́ᴥ•̀ʔっ ---> Debes de ingresar un número ")


if __name__ == '__main__':
    run()

Ingrese un número para obtener sus divisores: h

ʕ•́ᴥ•̀ʔっ ---> Debes de ingresar un número 


Ahora, utilizando una combinación de cláusulas `try`, `except` y `raise` tratemos de anticipar un error para cuando el usuario ingresa un número negativo.

Primero veamos que ocurre al ingresar un número negativo: 

In [40]:
def divisors(num):
    divisors = []
    for i in range (1, num + 1):
        if num % i == 0:
            divisors.append(i)
    return divisors

def run():
    num = int(input('Ingrese un número para obtener sus divisores: '))
    print(divisors(num))


if __name__ == '__main__':
    run()

Ingrese un número para obtener sus divisores: -1
[]


Podemos ver que no obtenemos un error como tal, pero para nada esperamos ese comportamiento en nuestro programa. Así que veamos como trabajar con esa excepción.

Como el programa se sigue ejecutando (aunque produciendo resultados incorrectos) sabemos que tendremos que hacer uso de `raise` para "elevar" la excepción de números negativos.

In [41]:
def divisors(num):
    divisors = []
    for i in range (1, num + 1):
        if num % i == 0:
            divisors.append(i)
    return divisors

def run():
    try:
        num = int(input('Ingrese un número para obtener sus divisores: '))
        
        if num < 0: 
            raise Exception("\nʕ•́ᴥ•̀ʔっ ---> No puedes ingresar números negativos")
        elif num == 0:
            raise Exception("\nʕ•́ᴥ•̀ʔっ ---> No puedes ingresar el número 0")
        
        print(divisors(num))
        
    except ValueError:
        print("\nʕ•́ᴥ•̀ʔっ ---> Debes de ingresar un número entero")
        
    except Exception as ve:
        print(ve)
        
    except Exception2 as ve2:
        print(ve2)
        
if __name__ == '__main__':
    run()

Ingrese un número para obtener sus divisores: -10

ʕ•́ᴥ•̀ʔっ ---> No puedes ingresar números negativos


Aquí agregamos los casos para cuando el tipo de datos es un `string` o `float` y que arroje el mensaje "Debes de ingresar un número entero". 

Después para el caso donde ingresamos números negativos la excepción devuelve "No puedes ingresar números negativos".

Y finalmente, para el caso donde se ingresa el 0 la excepción nos devuelve "No puedes ingresar el número 0"

---

### Assert statements

Hay una forma un poco más extravagante de manejar errores y que hasta cierto punto marca una diferencia en tu dominio del lenguaje y esos son los Assert statements o afirmaciones. Estas son expresiones (combinan variables con operadores) de forma que nos ayuda a definir si frente a nuestro programa estamos lidiando o no con un error.

Gráficamente lo podemos visualizar así:

![](https://i.imgur.com/OHjxkI6.png)

Como vemos en la imagen, el Assert statement nos ayuda a dirigir el flujo de nuestro programa, catalogándolo como error si es  o como una parte del código. Si la afirmación se cumple, devuelve `True` y seguimos con el código; pero si no se cumple, retorna `False` y manejamos el error.

La estructura básica de un assert sería la siguiente:

```Python
assert condición, mensaje de error
```

Donde nos ayudamos de la keyword `assert`. 

Se puede leer como, "afirmo que esta condición es verdadera, si es falsa retorna un mensaje de error".

Veamos un ejemplo:

In [123]:
def palindrome(string):
    return string == string[::-1]

print(palindrome(""))

True


Es el mismo caso que vimos anteriormente con el manejo de excepciones, pero ahora vamos a resolverlo con un `assert`. La implementación sería la siguiente:

In [137]:
def palindrome(string):
    assert len(string) > 0, "\nʕ•́ᴥ•̀ʔっ ---> No se puede ingresar una cadena vacía"
    return string == string[::-1]

print(palindrome(""))

AssertionError: 
ʕ•́ᴥ•̀ʔっ ---> No se puede ingresar una cadena vacía

Podemos ver que ahora obtenemos un mensaje con un error del tipo `AssertionError` es decir, una excepción basada en el `asser` que hicimos anteriormente.

Ahora vamos a resolver el mismo error de nuestro programa de divisores usando `assert statements`:

In [150]:
def divisors(num):
    divisors = []
    for i in range (1, num + 1):
        if num % i == 0:
            divisors.append(i)
    return divisors

def run():
    num = input('Ingrese un número para obtener sus divisores: ')
    assert num.isnumeric(), "\nʕ•́ᴥ•̀ʔっ ---> Debes ingresar un número."
    # Si la afirmación es verdadera pasalo a entero
    print(divisors(int(num)))


if __name__ == '__main__':
    run()

Ingrese un número para obtener sus divisores: asas


AssertionError: 
ʕ•́ᴥ•̀ʔっ ---> Debes ingresar un número.

Estamos haciendo uso de `.isnumeric()`, un método de los strings que nos permite saber si corresponde con un número o no. Así que asumimos que el input que estamos recibiendo es un número, si no lo es mandamos un mensaje de error.

Ahora, vamos a expandir nuestro alcance con Assert statements y usemos estos para evitar que se ingresen números negativos en el programa:

In [2]:
def divisors(num):
    divisors = []
    for i in range (1, num + 1):
        if num % i == 0:
            divisors.append(i)
    return divisors

def run():
    num = input('Ingrese un número para obtener sus divisores: ')
    
    assert num.isnumeric() and int(num) > 0, "\nʕ•́ᴥ•̀ʔっ ---> Debes ingresar un número positivo."
    # Si la afirmación es verdadera pasalo a entero
    print(divisors(int(num)))


if __name__ == '__main__':
    run()

Ingrese un número para obtener sus divisores: -1


AssertionError: 
ʕ•́ᴥ•̀ʔっ ---> Debes ingresar un número positivo.

Más información:

[Python Assert Statement - programiz.com](https://www.programiz.com/python-programming/assert-statement)

---

## Módulo 6. Manejo de archivos

### ¿Cómo trabajar con archivos?

Vamos a dividir los archivos que regularmente nos encontraremos en nuestra área de computación/programación en 2 categorías principales:

* Archivos de texto: Contiene bytes que representan letras y símbolos.

* Archivos binarios: Contienen bytes que representan cosas más complejas como vídeos, imágenes, etc.

![](https://i.imgur.com/jJLUVVU.png)

Para el caso donde utilizamos Python es común que nosotros casi no interactuemos con los archivos binarios (círculo verde). Nos concentraremos más en utilizar los archivos de texto.

Para abrir un archivo de texto con Python tenemos 3 formas de hacerlo:

* R $\rightarrow$ Lectura

* W $\rightarrow$ Escritura (sobrescribir)

* A $\rightarrow$ Escritura (agregar al final)

¿Cuál es el comando para abrir un archivo? 

```Python
with open("./ruta/del/archivo/archivo.txt , "r") as f:
```

La keyword `with` es un manejador contextual, es decir, lo que hace es controlar el flujo de nuestro archivo de forma que, si cerramos el programa o el script finaliza repentinamente este archivo no sufre daños. 

Después llamamos a la función de Python `open` que puede recibir varios parámetros, pero principalmente recibe 2:

* La ruta del archivo, en formato de string.
* Modo de apertura (r, w, a).

Y nuevamente usamos la keyword `as` la cual le asigna un nombre hipotético a nuestro archivo, en este caso `f`.

### Trabajando con archivos de texto en Python

Primero vamos a crear una carpeta llamada `archivos` en donde vamos a guardar un archivo de texto con información que usaremos en nuestro programa y le llamaremos `numbers.txt`, en donde agregaremos números aleatorios en cada una de las líneas. Y lo que queremos hacer es transformar a los datos de nuestro archivo en una lista de números.

Esto es lo que contiene nuestro archivo `../numbers.txt`:
    
![](https://i.imgur.com/xhoeB4q.png)

Vamos a crear un programa para entender como trabajar con archivos:

In [37]:
#Para leer
def read():
    numbers = []
    with open("../Programas/archivos/numbers.txt" , "r", encoding="UTF-8") as  f:
        for line in f: 
            numbers.append(int(line))
        print(numbers)

  
#Para sobrescribir
def write_w_mode():
    names = ["Abdiel", "uwu", ":(", "Holi", "Jesús"]
    with open("../Programas/archivos/names.txt", "w", encoding="UTF-8") as f:
        for name in names:
            f.write(name)
            f.write("\n")
 

#Para escribir al final
def write_a_mode():
    names = ["Aquí", "los", "nuevos", "Nombres", ":)"]
    with open("../Programas/archivos/names.txt", "a", encoding="UTF-8") as f:
        for name in names:
            f.write(name)
            f.write("\n")
            
def run():
    read()
    #write_w_mode()
    #write_a_mode()


if __name__ == '__main__':
    run()

[1, 45, 67, 34, 21, 124, 545, 324, 7678, 87687, 2232434, 3, 212]


Para indicar la ruta relativa de un archivo en Windows usamos `../` antes de escribir el resto de la ruta.

El parámetro `encoding="UTF-8"` dentro de `open()` nos permite leer o escribir sin problemas cuando trabajamos con archivos que contienen caracteres especiales.

También podemos llamar al método `.write()` de nuestro archivo `f` para escribir sobre dicho archivo.

Nuestra primera función `read()`, se encargará únicamente de *leer el archivo* y a partir de lo que lea en cada una de las líneas convertirá ese valor a un entero y lo agregará a una lista y la imprimimos, de ahí obtenemos la lista `[1, 45, 67, 34, 21, 124, 545, 324, 7678, 87687, 2232434, 3, 212]`. 

Después, en nuestra función `write_w_mode()` vamos a crear una lista con algunos strings y después vamos a crear un archivo llamado `../names.txt` y agregaremos cada uno de los strings en dicho archivo, sobrescribiendo el archivo cada que llamemos a la función ya que tiene el parámetro `"w"`. Obteniendo el siguiente archivo:

![](https://i.imgur.com/A1icQrj.png)

Finalmente con nuestra función `write_a_mode()` haremos algo muy similar a ls función anterior, pero al tener el parámetro `"a"` añadirá los nuevos nombres al final del archivo. Es decir, en lugar de sobrescribir con un nuevo archivo solo lo completará con la nueva información.

Obteniendo el siguiente archivo:

![](https://i.imgur.com/mW7eiAN.png)

---

## Conclusiones.

### Programa final: Hangman Game

Vamos a hacer un pequeño programa que simule el Hangman Game. Y tendrá que hacer lo siguiente:

* Al ejecutar el programa se tendrá que limpiar la pantalla y mostrar algo así:

![](https://i.imgur.com/UuYPz9E.png)

* Al ingresar una letra, si coincide con una de las letras que contiene la palabra se tendrá que actualizar y mostrar en donde está esa letra, por ejemplo si presionamos la a:

![](https://i.imgur.com/dEmtd4F.png)

* En caso de ingresar una letra incorrecta simplemente te arrojará de nuevo el mensaje de que ingreses una nueva letra hasta que demos con la palabra correcta.

* Una vez adivinando la palabra se obtiene un mensaje como este:

![](https://i.imgur.com/OdQJD8g.png)

Todo esto siguiendo unas cuantas reglas:

* Incorporando comprehensions, manejo de errores y manejo de archivos.

* Utiliza el archivo `data.txt` y léelo para obtener las palabras.

#### Ayudas y pistas:

* Investiga la función `enumerate`.

* El método `get` de los diccionarios puede servir.

* La sentencia <br>
    `os.system("cls")` $\rightarrow$ Windows <br>
    `os.system("clear")` $\rightarrow$ Unix <br>
    que sirve para limpiar la pantalla después de que ingresamos una letra.

#### Posibles mejoras al juego:

* Añade un sistema de puntos.

* Dibuja un "ahorcado" en cada jugada con código ASCII.

* Mejora la interfaz.

#### Código:

In [39]:
import random
import os
from functools import reduce

drawing = [''' 
 _     _                                               ______                      
| |   | |                                             / _____)                     
| |__ | |  ____  ____    ____  ____    ____  ____    | /  ___   ____  ____    ____ 
|  __)| | / _  ||  _ \  / _  ||    \  / _  ||  _ \   | | (___) / _  ||    \  / _  )
| |   | |( ( | || | | |( ( | || | | |( ( | || | | |  | \____/|( ( | || | | |( (/ / 
|_|   |_| \_||_||_| |_| \_|| ||_|_|_| \_||_||_| |_|   \_____/  \_||_||_|_|_| \____)
                       (_____|                                                     
                                                                                        
''',
'''
   +---+
   |   |
       |
       |
       |
       |
=========''', 
           
'''
   +---+
   |   |
   O   |
       |
       |
       |
=========''', 
           
'''
   +---+
   |   |
   O   |
   |   |
       |
       |
 =========''', 
                     
'''
   +---+
   |   |
   O   |
  /|   |
       |
       |
 =========''',
                     
'''
   +---+
   |   |
   O   |
  /|\  |
       |
       |
 =========''',
                     
'''
   +---+
   |   |
   O   |
  /|\  |
  /    |
       |
 =========''', 
                     
'''
   +---+
   |   |
   O   |
  /|\  |
  / \  |
       |
=========''']

# Obtenemos una palabra random de nuestro archivo de palabras
def getting_word():
    words = []
    with open("../Programas/archivos/diccionario_palabras.txt" , "r", encoding="UTF-8") as  f:
        for line in f:
            # removemos espacios y ponemos todo en mayusculas
            words.append(line.strip().upper())
    word = random.choice(words)
    print(word)
    #print(len(word))
    return word

# Omitimos los acentos
def without_accent_mark(word):
    word_changed = (
        ("Á", "A"),
        ("É","E"),
        ("Í","I"),
        ("Ó","O"),
        ("Ú","U"),
    ) 
    for a,b in word_changed:
        word = word.replace(a,b)
    print(word)
    return word 

# Hacemos el dibujito de la interfaz
def create_interfaz(changed):
    underscore = [('_') for n in changed]
    #print(underscore)
    underscores = reduce(lambda a,b: a+(" ")+b,underscore)
    print(drawing[0])
    print("¡Adivina la palabra!")
    print(drawing[1])
    print("\n", underscores)
    return underscores,changed

def run_name(underscore):
    vidas = 5
    entrada = input('\nIngresa una letra:').upper()
    print(entrada)
    print(underscore)
    for i in underscore[1]:
        print(i)
        if entrada == i:
            print('Letra correcta')
    
def run():
    word = getting_word()
    changed = without_accent_mark(word)
    underscore = create_interfaz(changed)
    run_name(underscore)

if __name__ == '__main__':
    run()

INSTITUTORA
INSTITUTORA
 
 _     _                                               ______                      
| |   | |                                             / _____)                     
| |__ | |  ____  ____    ____  ____    ____  ____    | /  ___   ____  ____    ____ 
|  __)| | / _  ||  _ \  / _  ||    \  / _  ||  _ \   | | (___) / _  ||    \  / _  )
| |   | |( ( | || | | |( ( | || | | |( ( | || | | |  | \____/|( ( | || | | |( (/ / 
|_|   |_| \_||_||_| |_| \_|| ||_|_|_| \_||_||_| |_|   \_____/  \_||_||_|_|_| \____)
                       (_____|                                                     
                                                                                        

¡Adivina la palabra!

   +---+
   |   |
       |
       |
       |
       |

 _ _ _ _ _ _ _ _ _ _ _

Ingresa una letra:f
F
('_ _ _ _ _ _ _ _ _ _ _', 'INSTITUTORA')
I
N
S
T
I
T
U
T
O
R
A
